### ISSUE TITLE
How can conditional appending be implemented without depending on the single writer pattern?

### ISSUE THREAD

**Issue by Narvalex** [m1l4n54v1c/event-store#3]
One of the most pressing questions I've come across, and that I also have, concerns the implementation details of conditional append. It seems to require a serializable transaction isolation level in RDBMS, and in document databases, it appears to need a single writer pattern, which could slow things down compared to the typical optimistic concurrency checks that most databases support. I understand that some internal Axon Server prototypes might not be shareable, but it would be great to get at least a hint about this topic. 😉

**Comment by gklijs** [comment]
Why do you need a single writer? As long as any writer can check just before persisting the writes if none of the other writers wrote something that might break the consistency boundary? If so, take those additional events into account and try again.

**Comment by Narvalex** [comment]
> Why do you need a single writer? As long as any writer can check just before persisting the writes if none of the other writers wrote something that might break the consistency boundary? If so, take those additional events into account and try again.

While the idea of allowing multiple writers to check for conflicts before persisting changes sounds reasonable, it introduces a potential race condition. If two writers check for consistency at the same time, they won't be aware of each other's pending writes. Both could then assume it's safe to proceed, leading to both of them persisting changes concurrently. This results in conflicting updates being committed, which could push the system into an inconsistent state.

By contrast, a single writer guarantees that no two changes are being committed at the same time, preventing these kinds of conflicts from occurring.

**Comment by gklijs** [comment]
No, one of them is persisted, the other isn't but are cleaned. On the next try the new events are taking into consideration, and the events are persisted. If there are a lot of race conditions like this, it might have a high cost, but it should be possible to get to an inconsistent state if well implemented.

**Comment by Narvalex** [comment]
> No, one of them is persisted, the other isn't but are cleaned. On the next try the new events are taking into consideration, and the events are persisted. If there are a lot of race conditions like this, it might have a high cost, but it should be possible to get to an inconsistent state if well implemented.

How would the second writer know that it needs to roll back? What mechanism would it use to detect that the data has changed since the read, or that there is a conflict with another writer?

**Comment by gklijs** [comment]
A simple quick check on what was written and persisted between when it read the events till now. The check can easily be part of the transaction, thus excluding race conditions that might lead to an inconsistent state.

**Comment by Narvalex** [comment]
I think is not simple. In databases with snapshot isolation, like Cosmos DB or MongoDB, reads within the same transaction do not reflect concurrent writes from other transactions, making it challenging to detect such changes and prevent race conditions.

**Comment by gklijs** [comment]
I know it is simple. But you shouldn't implement it on top of something like MongoDB. I used an embedded database in some prototype. This does open all sorts of other interesting things like backups, consensus among nodes, performance and metrics.
